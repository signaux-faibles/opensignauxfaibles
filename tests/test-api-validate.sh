#!/bin/bash

# Test de bout en bout de POST /api/data/validate.
# Inspir√© de test-api-export.sh.
# Ce script doit √™tre ex√©cut√© depuis la racine du projet. Ex: par test-all.sh.

tests/helpers/mongodb-container.sh stop

set -e # will stop the script if any command fails with a non-zero exit code

# Setup
FLAGS="$*" # the script will update the golden file if "--update" flag was provided as 1st argument
INPUT_FILE="tests/input-data/RawData.validation.json"
GOLDEN_FILE="tests/output-snapshots/test-api-validate.golden.json"
TMP_DIR="tests/tmp-test-execution-files"
OUTPUT_FILE="${TMP_DIR}/output.json"
mkdir -p "${TMP_DIR}"

# Clean up on exit
function teardown {
    tests/helpers/dbmongo-server.sh stop || true # keep tearing down, even if "No matching processes belonging to you were found"
    tests/helpers/mongodb-container.sh stop
}
trap teardown EXIT

PORT="27016" tests/helpers/mongodb-container.sh start

MONGODB_PORT="27016" tests/helpers/dbmongo-server.sh setup

echo ""
echo "üìù Inserting test data..."
sleep 1 # give some time for MongoDB to start
tests/helpers/mongodb-container.sh run << CONTENT
  db.RawData.insertMany($(cat ${INPUT_FILE}))
CONTENT

echo ""
echo "üíé Testing the dbmongo API..."
tests/helpers/dbmongo-server.sh start
API_RESULT=$(http --print=b --ignore-stdin :5000/api/data/validate collection=RawData)
echo "- POST /api/data/validate üëâ ${API_RESULT}"

(tests/helpers/mongodb-container.sh run \
  > "${OUTPUT_FILE}" \
) << CONTENT
print("// db.Journal:");
const report = db.Journal.find().toArray().pop() || {};
printjson({
  count: db.Journal.count(),
  reportType: report.reportType,
  hasDate: !!report.date,
  hasStartDate: !!report.startDate,
});

print("// Result from /api/data/validate:");
CONTENT

OUTPUT_GZ_FILE=dbmongo/$(echo ${API_RESULT} | tr -d '"')
zcat < "${OUTPUT_GZ_FILE}" >> "${OUTPUT_FILE}"

tests/helpers/diff-or-update-golden-master.sh "${FLAGS}" "${GOLDEN_FILE}" "${OUTPUT_FILE}"

rm "${OUTPUT_GZ_FILE}"
rm -rf "${TMP_DIR}"
# Now, the "trap" commands will clean up the rest.
