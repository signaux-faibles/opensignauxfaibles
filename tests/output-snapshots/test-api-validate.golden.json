// db.Journal:
{
	"count" : 1,
	"reportType" : "ValidateDataEntries",
	"hasDate" : true,
	"hasStartDate" : true
}
// Result from /api/data/validate:
{"_id":"01234567891011","batchKey":"1802","dataHash":"hash-d√©lai-nul","dataObject":{"date_creation":"2016-09-01T00:00:00Z","date_echeance":"2016-09-01T00:00:00Z","duree_delai":0,"montant_echeancier":10},"dataType":"delai"}
{"_id":"01234567891011","batchKey":"1803","dataHash":"hash-siren-manquant","dataObject":{},"dataType":"bdf"}
{"_id":"01234567891011","batchKey":"1803","dataHash":"hash-siren-invalide","dataObject":{"siren":"123"},"dataType":"bdf"}
{"_id":"01234567891011","batchKey":"1803","dataHash":"hash-date-manquante","dataObject":{"date_echeance":"2016-09-01T00:00:00Z","duree_delai":1,"montant_echeancier":10},"dataType":"delai"}
{"_id":"01234567891011","batchKey":"1803","dataHash":"hash-montant-nul","dataObject":{"date_creation":"2016-09-01T00:00:00Z","date_echeance":"2016-09-01T00:00:00Z","duree_delai":1,"montant_echeancier":0},"dataType":"delai"}
